### PR TITLE
DSND-2170 Implement filing history PUT request for TM01

### PR DESCRIPTION
**Pull request by cgregoryCH**
### Related Jira tickets

[DSND-2170](https://companieshouse.atlassian.net/browse/DSND-2170)

### General

- [x] Is the PR as small as it can be?
- [x] Has the Jira ticket been updated with any relevant comments?
- [ ] Is the `README` up to date?
- [ ] Has the `POM` been updated for the latest versions of dependencies?
- [x] Has the code been double-checked against `main`?
- [ ] Does the code adhere standards in this repository, including SonarQube checks?

### Testing

- [ ] Do the code changes have unit and integration tests with code coverage > 80%?
- [ ] Has the code been tested locally and/or passed the API Karate tests on Tilt?
- [ ] Have mandatory manual test cases been run?
- [ ] Are extra Karate tests required?
- [ ] Are all the test data resources up to date?

### Release preparation

_Where possible, add links to the respective Jira tickets when an item is checked_

- [ ] Have these changes been included in a release ticket?
- [ ] Are changes required to the `docker-chs-development` deployment or test data?
- [ ] Are any changes required to the environment added to deployment repositories?

## Notes to the tester

_Add any testing hints or instructions here._

[DSND-2170]: https://companieshouse.atlassian.net/browse/DSND-2170?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Review by kinpang-CH — APPROVED** [review]

File: pom.xml [hunk 1]
```diff
@@ -80,6 +92,16 @@
       <artifactId>grpc-context</artifactId>
       <version>${grpc.version}</version>
     </dependency>
+    <dependency>
+      <groupId>org.springframework</groupId>
+      <artifactId>spring-core</artifactId>
```
Comment (johncook-ch) [inline pom.xml:97]: Should this really be brought in from the Spring Boot BOM? I am concerned about mixing core Spring modules from different release versions

File: src/main/java/uk/gov/companieshouse/filinghistory/api/controller/FilingHistoryController.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+        DataMapHolder.get()
+                .companyNumber(companyNumber)
+                .transactionId(transactionId);
+        LOGGER.info("Processing transaction upsert", DataMapHolder.getLogMap());
+
+        final ServiceResult result = serviceProcessor.processFilingHistory(transactionId, requestBody);
+
+        // This is a switch because we'll need to add more cases in the future when doing unhappy paths
+        return switch (result) {
```
Comment (johncook-ch) [inline src/main/java/uk/gov/companieshouse/filinghistory/api/controller/FilingHistoryController.java:43]: Nice :)

File: src/main/java/uk/gov/companieshouse/filinghistory/api/model/FilingHistoryData.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
+package uk.gov.companieshouse.filinghistory.api.model;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Objects;
+import org.springframework.data.mongodb.core.mapping.Field;
+
+public class FilingHistoryData {
```
Comment (johncook-ch) [inline src/main/java/uk/gov/companieshouse/filinghistory/api/model/FilingHistoryData.java:8]: Could this be a `record` instead?

File: src/test/java/uk/gov/companieshouse/filinghistory/api/controller/FilingHistoryControllerIT.java [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+import uk.gov.companieshouse.filinghistory.api.model.FilingHistoryData;
+import uk.gov.companieshouse.filinghistory.api.model.FilingHistoryDescriptionValues;
+import uk.gov.companieshouse.filinghistory.api.model.FilingHistoryDocument;
+import uk.gov.companieshouse.filinghistory.api.model.FilingHistoryLinks;
+import uk.gov.companieshouse.filinghistory.api.model.FilingHistoryOriginalValues;
+
+@Testcontainers
+@AutoConfigureMockMvc
+@SpringBootTest(classes = FilingHistoryApplication.class, webEnvironment = SpringBootTest.WebEnvironment.MOCK)
+class FilingHistoryControllerIT {
```
Comment (johncook-ch) [inline src/test/java/uk/gov/companieshouse/filinghistory/api/controller/FilingHistoryControllerIT.java:48]: Can this class also test for auth failures etc?

**Review by johncook-ch — CHANGES_REQUESTED** [review]

File: src/test/java/uk/gov/companieshouse/filinghistory/api/controller/FilingHistoryControllerIT.java (hunk 4 above)
Comment (cgregoryCH) [reply src/test/java/uk/gov/companieshouse/filinghistory/api/controller/FilingHistoryControllerIT.java:48]: Sam and I decided they'd be done in the un-happy path ticket, is that okay?

File: src/main/java/uk/gov/companieshouse/filinghistory/api/model/FilingHistoryData.java (hunk 3 above)
Comment (cgregoryCH) [reply src/main/java/uk/gov/companieshouse/filinghistory/api/model/FilingHistoryData.java:8]: We would need to create a builder within it as the unit tests build partial objects. Would that be preferable?

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/test/java/uk/gov/companieshouse/filinghistory/api/controller/FilingHistoryControllerIT.java (hunk 4 above)
Comment (johncook-ch) [reply src/test/java/uk/gov/companieshouse/filinghistory/api/controller/FilingHistoryControllerIT.java:48]: Yes. I forgot that we had split those out

File: src/main/java/uk/gov/companieshouse/filinghistory/api/model/FilingHistoryData.java (hunk 3 above)
Comment (johncook-ch) [reply src/main/java/uk/gov/companieshouse/filinghistory/api/model/FilingHistoryData.java:8]: Let's leave it for now and revit. Adding a Builder to a record is simple

**Pull request merged**